src/pandas_openscm/index_manipulation.py

```diff
@@ -714,3 +717,188 @@ def update_levels_from_other(
+    ...     start,
+    ...     {"new_variable": [1, 2, 3, 4]},
+    ... )
+    MultiIndex([('sa', 'ma', 'v1', 'kg', 1),
+                ('sb', 'ma', 'v2',  'm', 2),
+                ('sa', 'mb', 'v1', 'kg', 3),
+                ('sa', 'mb', 'v2',  'm', 4)],
+               names=['scenario', 'model', 'variable', 'unit', 'new_variable'])
+    >>>
+    >>> # Replace a level with a single value and add a new level
+    >>> set_levels(
+    ...     start,
+    ...     {"model": "new_model", "new_variable": ["xyz", "xyz", "x", "y"]},
+    ... )
+    MultiIndex([('sa', 'new_model', 'v1', 'kg', 'xyz'),
+                ('sb', 'new_model', 'v2',  'm', 'xyz'),
+                ('sa', 'new_model', 'v1', 'kg',   'x'),
+                ('sa', 'new_model', 'v2',  'm',   'y')],
+               names=['scenario', 'model', 'variable', 'unit', 'new_variable'])
+    """
+    levels: list[pd.Index[Any]] = list(ini.levels)
+    codes: list[list[int] | npt.NDArray[np.integer[Any]]] = list(ini.codes)
+    names: list[str] = list(ini.names)
+
+    for level, value in levels_to_set.items():
+        if isinstance(value, Collection) and not isinstance(value, str):
+            if len(value) != len(ini):
+                msg = (
+                    f"Length of values for level '{level}' does not "
+                    f"match index length: {len(value)} != {len(ini)}"
+                )
+                raise ValueError(msg)
+            new_level, new_codes = create_level_from_collection(level, value)
+        else:
+            new_level = pd.Index([value], name=level)
+            new_codes = [0] * ini.shape[0]
+
+        if level in ini.names:
+            level_idx = ini.names.index(level)
+            levels[level_idx] = new_level
+            codes[level_idx] = new_codes
+        else:
+            levels.append(new_level)
+            codes.append(new_codes)
+            names.append(level)
+
+    res = pd.MultiIndex(levels=levels, codes=codes, names=names)
+
+    return res
+
+
+def set_index_levels_func(
+    df: pd.DataFrame,
+    levels_to_set: dict[str, Any | Collection[Any]],
+    copy: bool = True,
+) -> pd.DataFrame:
+    """
+    Set the index levels of a [pd.DataFrame][pandas.DataFrame]
+
+    Parameters
+    ----------
+    df
+        [pd.DataFrame][pandas.DataFrame] to update
+
+    levels_to_set
+        Mapping of level names to values to set
+
+    copy
+        Should `df` be copied before returning?
+
+
+    Returns
+    -------
+    :
+        `df` with updates applied to its index
+    """
+    if not isinstance(df.index, pd.MultiIndex):
+        msg = f"Expected MultiIndex, got {type(df.index)}"
+        raise TypeError(msg)
+
+    if copy:
+        df = df.copy()
+
+    if not isinstance(df.index, pd.MultiIndex):
+        msg = (
+            "This function is only intended to be used "
+            "when `df`'s index is an instance of `MultiIndex`. "
+            f"Received {type(df.index)=}"
+        )
+        raise TypeError(msg)
```

Comment:
```suggestion
    if not isinstance(df.index, pd.MultiIndex):
        msg = (
            "This function is only intended to be used "
            "when `df`'s index is an instance of `MultiIndex`. "
            f"Received {type(df.index)=}"
        )
        raise TypeError(msg)

    if copy:
        df = df.copy()
```

Tidying up so we only do the check for being a `pd.MultiIndex` once (codecov picked this one up, which is a good illustration of a rare case where a code coverage checker is actually quite helpful)